jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/Retainable.java

```diff
@@ -13,7 +13,139 @@
 
 package org.eclipse.jetty.util;
 
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * <p>A reference counted resource, for example one that is borrowed from a pool,
+ * that may be retained an additional number of times, and released a correspondent
+ * number of times, over its lifecycle.</p>
+ * <p>The resource is typically implicitly retained when it is first created.
+ * It may be retained more times (thus incrementing its reference count) and released
+ * (thus decrementing its reference count), until the reference count goes to zero.</p>
+ */
 public interface Retainable
 {
-    public void retain();
+    /**
+     * <p>Retains this resource, incrementing the reference count.</p>
+     */
+    public default void retain()
+    {
+    }
```

Comment:
I find an empty method here a little bit dangerous.  You have several implementions that implement release to do something, but there is no actual reference count, so calls to retain are ignored.

If you want a default, then it should be:
```suggestion
    {
        throw new UnsupportedOperationException();
    }
```

But it is probably better to force implementations to implement this method, so that only the ones that have a noop release also have a noop retain.

At the very least, you need javadoc that says implementations that implement release, must implement retain and respect the contract.

Reply:
I made `Retainable` methods abstract now, so every implementation must implement them.
However, `release()` cannot throw if `retain()` is empty, because special implementations such as `Chunk.Error` or `Chunk.EOF` must have a release method that does not throw.
However, should it return `true` or `false`? For now I decided for `false` (contrary to the previous implementation that returned `true`.
Typically callers don't care about the return value of `release()`, only wrappers do, so I guess either `false` or `true` is pretty much the same, but I now have a slight preference for `false`.

Reply:
Ok I changed my mind. I now prefer to throw `UnsupportedOperationException` from `retain()` and return `true` from `release()` for special implementations.